src/base.rs

```diff
@@ -466,63 +422,200 @@ impl fmt::Display for Location {
 /// A piece of assertion message.
 ///
 /// # Design discussion
-/// - New entry for having elements?
-///     - `KeyValues {key: String, value: Vec<String>}`
 /// - New entry for comparing elements?
 ///     - `Comparison {key: String, actual: Vec<String>, expected: Vec<String>}`
 #[allow(missing_docs)]
 #[derive(Debug, Clone, Eq, PartialEq)]
 pub enum Fact {
-    /// Keyed assertion message
-    ///
-    /// # Example
-    /// ```text
-    /// Fact {key: "expected", value: "foo"}
-    /// Fact {key: "actual", value: "var"}
-    /// ```
-    KeyValue { key: String, value: String },
-    /// Keyed assertion message for multiple values
-    ///
-    /// # Example
-    /// ```text
-    /// Fact {key: "expected", values: vec!["foo", "bar"]}
-    /// ```
-    KeyValues { key: String, values: Vec<String> },
-    /// Single assertion message
-    ///
-    /// # Example
-    /// ```text
-    /// Fact {value: "expected that the vec is empty"}
-    /// ```
-    Value { value: String },
+    Structural {
+        inner: FactStructure,
+    },
     /// Splitter
     Splitter,
 }
 
 #[allow(missing_docs)]
 impl Fact {
     pub fn new<K: Into<String>, V: Into<String>>(key: K, value: V) -> Fact {
-        Fact::KeyValue {
-            key: key.into(),
-            value: value.into(),
+        Fact::Structural {
+            inner: FactStructure::KeyValue {
+                key: key.into(),
+                value: Box::new(FactStructure::Value {
+                    formatted_value: value.into(),
+                }),
+            },
         }
     }
     pub fn new_simple_fact<V: Into<String>>(value: V) -> Fact {
-        Fact::Value {
-            value: value.into(),
+        Fact::Structural {
+            inner: FactStructure::Value {
+                formatted_value: value.into(),
+            },
         }
     }
     pub fn new_multi_value_fact<K: Into<String>, V: Into<String>>(key: K, values: Vec<V>) -> Fact {
-        Fact::KeyValues {
-            key: key.into(),
-            values: values.into_iter().map(|v| v.into()).collect(),
+        Fact::Structural {
+            inner: FactStructure::KeyValue {
+                key: key.into(),
+                value: Box::new(FactStructure::List {
+                    values: values
+                        .into_iter()
+                        .map(|v| {
+                            Box::new(FactStructure::Value {
+                                formatted_value: v.into(),
+                            })
+                        })
+                        .collect(),
+                }),
+            },
         }
     }
     pub fn new_splitter() -> Fact {
         Fact::Splitter
     }
 }
 
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum FactStructure {
```

Comment:
I guess you defined a new enum structure to treat splitter differently though, defining two enum has some downsides like writing tests gets complicated because every meaningful fact gets nested due to wrapping.

I'm leaning to think that one enum may better.  WDYT?